pkg/sharings/recipients.go

```diff
@@ -57,46 +70,104 @@ func (r *Recipient) ExtractDomain() (string, error) {
 	return r.URL, nil
 }
 
-// GetAccessToken sends an access_token requests to the recipient, using the
-// given authorization code
-func (r *Recipient) GetAccessToken(code string) (*auth.AccessToken, error) {
-	client := new(http.Client)
-	rURL, err := r.ExtractDomain()
+// CreateRecipient inserts a Recipient document in database
+func CreateRecipient(db couchdb.Database, doc *Recipient) error {
+	// Sanity check: we need an URL…
+	if doc.Email == "" {
+		return ErrRecipientHasNoEmail
+	}
+	// …and an E-mail.
+	if doc.URL == "" {
+		return ErrRecipientHasNoURL
+	}
+
+	err := couchdb.CreateDoc(db, doc)
+	return err
+}
+
+// GetRecipient returns the Recipient stored in database from a given ID
+func GetRecipient(db couchdb.Database, recID string) (*Recipient, error) {
+	doc := &Recipient{}
+	err := couchdb.GetDoc(db, consts.Recipients, recID, doc)
+	if couchdb.IsNotFoundError(err) {
+		err = ErrRecipientDoesNotExist
+	}
+	return doc, err
+}
+
+// getAccessToken sends an "access_token" request to the recipient using the
+// given authorization code.
+func (rs *RecipientStatus) getAccessToken(db couchdb.Database, code string) (*auth.AccessToken, error) {
+	// Sanity check: `recipient` being a private attribute it can be nil if the
+	// sharing document that references it was extracted from the database.
+	if rs.recipient == nil {
+		recipient, err := GetRecipient(db, rs.RefRecipient.ID)
+		if err != nil {
+			return nil, err
+		}
+		rs.recipient = recipient
+	}
+
+	// Other sanity check: the recipient has to have an URL and an email.
+	if rs.recipient.URL == "" {
+		return nil, ErrRecipientHasNoURL
+	}
+	if rs.Client.ClientID == "" {
+		return nil, ErrNoOAuthClient
+	}
+
+	// The structure `auth.Request` expects a domain WITHOUT a scheme (i.e.
+	// without "http://" or "https://") so we parse it.
+	recipientURL, err := rs.recipient.ExtractDomain()
```

Comment:
You can call it a `recipientDomain` instead of `recipientURL` if it is just the domain